src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -191,8 +191,32 @@ private async Task PullContainerAsync(IExecutionContext executionContext, Contai
                 ArgUtil.NotNullOrEmpty(username, nameof(username));
                 ArgUtil.NotNullOrEmpty(password, nameof(password));
 
-                int loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
-                if (loginExitCode != 0)
+
+                // login with retry up to 3 times
+                int maxRetries = 3;
+                int retryCount = 0;
+                int loginExitCode = 0;
+
+                while (retryCount < maxRetries)
+                {
+                    loginExitCode = await _dockerManger.DockerLogin(executionContext, registryServer, username, password);
+                    if (loginExitCode == 0)
+                    {
+                        break;
+                    }
+                    else
+                    {
+                        retryCount++;
+                        if (retryCount < maxRetries)
+                        {
+                            var backOff = BackoffTimerHelper.GetRandomBackoff(TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(10));
```

Comment:
Suggestion: it is more preferrable to use [GetExponentialBackoff](https://learn.microsoft.com/en-us/previous-versions/dn242425(v=vs.120)) instead of GetRandomBackoff

Reply:
Done!